File: main/java/ee/schimke/oksocial/output/OutputHandler.java
package ee.schimke.oksocial.output;

import java.io.IOException;

public interface OutputHandler<R> {
  void showOutput(R response) throws IOException;

  default void showError(String message, Throwable e) {
    if (message != null) {
      System.err.println(message);
    }
    if (e != null) {
      e.printStackTrace();
    }
  }

  default void openLink(String url) throws IOException {
    System.err.println(url);
  }

  default void info(String message) {
    System.out.println(message);
  }
}

File: src/main/kotlin/com/baulsupp/oksocial/output/OutputHandler.kt
package com.baulsupp.oksocial.output

/**
 * Interface for command line apps to output content.
 */
interface OutputHandler<in R> {
  suspend fun showOutput(response: R)

  suspend fun showError(message: String? = null, e: Throwable? = null) {
    if (message != null) {
      System.err.println(message)
    }
    e?.printStackTrace()
  }

  suspend fun openLink(url: String) {
    System.err.println(url)
  }

  fun info(message: String) {
    println(message)
  }
}
